### PR TITLE
dependency: bump Saxon-HE from 9.9.1-7 to 10.0

### PR DESCRIPTION
**Pull request by HuGanghui**
Bumps Saxon-HE from 9.9.1-7 to 10.0.

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

### CODE REVIEW

**Comment by HuGanghui** [comment]
```java
@Test
    public void testPrintXpathNodeParentNull(@SysErr Capturable systemErr,
            @SysOut Capturable systemOut) throws IOException {
        final String expected = addEndOfLine("PACKAGE_DEF -> package [1:0]");
        Main.main("-b", "/PACKAGE_DEF", getPath("InputMainXPath.java"));
        assertThat("Unexpected output log", systemOut.getCapturedData(), is(expected));
        assertThat("Unexpected system error log", systemErr.getCapturedData(), is(""));
    }
```

**Comment by HuGanghui** [comment]
@timurt many tests failed because of this update Saxon-HE from 9.9.1-7 to 10.0.  can you give some advice how to solve these tests?

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java [hunk 1]
```diff
@@ -201,10 +202,11 @@ public NodeInfo getRoot() {
      * when there is no axis iterator for given axisNumber.
      *
      * @param axisNumber element from {@code AxisInfo}
+     * @param nodeTest filter for iterator
      * @return {@code AxisIterator} object
      */
     @Override
-    public AxisIterator iterateAxis(byte axisNumber) {
+    public AxisIterator iterateAxis(int axisNumber, Predicate<? super NodeInfo> nodeTest) {
```
Comment (timurt) [inline src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java:209]: replace return statement with `return new Navigator.AxisFilter(result, nodeTest);`

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/RootNode.java [hunk 2]
```diff
@@ -142,10 +144,11 @@ public NodeInfo getRoot() {
      * when there is no axis iterator for given axisNumber.
      *
      * @param axisNumber element from {@code AxisInfo}
+     * @param nodeTest filter for iterator
      * @return {@code AxisIterator} object
      */
     @Override
-    public AxisIterator iterateAxis(byte axisNumber) {
+    public AxisIterator iterateAxis(int axisNumber, Predicate<? super NodeInfo> nodeTest) {
```
Comment (timurt) [inline src/main/java/com/puppycrawl/tools/checkstyle/xpath/RootNode.java:151]: replace return statement with `return new Navigator.AxisFilter(result, nodeTest);`

**Review by timurt — CHANGES_REQUESTED** [review]

**Comment by timurt** [comment]
@HuGanghui I think I found problem. In previous versions filtering results were done automatically by library itself, now it is on user side ([see doc here](https://www.saxonica.com/html/documentation/javadoc/net/sf/saxon/om/NodeInfo.html#iterateAxis-int-java.util.function.Predicate-))

please update `iterateAxis` implementations for `ElementNode` and `RootNode` as I suggested and fix unit tests

**Comment by HuGanghui** [comment]
@timurt ok, I think your idea is right, and after `replace return statement with return new Navigator.AxisFilter(result, nodeTest);`,   test errors I think are caused by the same reason.

a error example:

```
@Test
    public void testRootWithNullDetailAst() {
        final RootNode emptyRootNode = new RootNode(null);
        assertThat("Empty node should not have children", emptyRootNode.hasChildNodes(),
                equalTo(false));
        assertThat("Invalid number of nodes", emptyRootNode.iterateAxis(AxisInfo.DESCENDANT),
                equalTo(EmptyIterator.ofNodes()));
        assertThat("Invalid number of nodes", emptyRootNode.iterateAxis(AxisInfo.CHILD),
                equalTo(EmptyIterator.ofNodes()));
    }
```

--->

```
@Test
    public void testRootWithNullDetailAst() {
        final RootNode emptyRootNode = new RootNode(null);
        final Predicate<? super NodeInfo> nodeTest = null;
        assertThat("Empty node should not have children", emptyRootNode.hasChildNodes(),
                equalTo(false));
        assertThat("Invalid number of nodes", emptyRootNode.iterateAxis(AxisInfo.DESCENDANT),
                equalTo(Navigator.AxisFilter(EmptyIterator.ofNodes(), nodeTest)));
        assertThat("Invalid number of nodes", emptyRootNode.iterateAxis(AxisInfo.CHILD),
                equalTo(EmptyIterator.ofNodes()));
    }
```

[ERROR] testRootWithNullDetailAst  Time elapsed: 0.012 s  <<< FAILURE!
java.lang.AssertionError: 
Invalid number of nodes
Expected: <net.sf.saxon.tree.iter.EmptyIterator$OfNodes@73167877>
     but: was <net.sf.saxon.tree.util.Navigator$AxisFilter@243eb8d5>
        at com.puppycrawl.tools.checkstyle.xpath.XpathMapperTest.testRootWithNullDetailAst(XpathMapperTest.java:485)

**Comment by timurt** [comment]
@HuGanghui but why you rewrote only first assert 
```
@Test
    public void testRootWithNullDetailAst() {
        final RootNode emptyRootNode = new RootNode(null);
        final Predicate<? super NodeInfo> nodeTest = null;
        assertThat("Empty node should not have children", emptyRootNode.hasChildNodes(),
                equalTo(false));
        assertThat("Invalid number of nodes", emptyRootNode.iterateAxis(AxisInfo.DESCENDANT),
                equalTo(Navigator.AxisFilter(EmptyIterator.ofNodes(), nodeTest)));
        assertThat("Invalid number of nodes", emptyRootNode.iterateAxis(AxisInfo.CHILD),
                equalTo(EmptyIterator.ofNodes())); // this one also should be rewritten
    }
```

There are several unit tests in `ElementNodeTest`, `RootNodeTest` and `XpathMapperTest` those are expecting specific type of iterator, these tests should be also fixed.

**Comment by HuGanghui** [comment]
@timurt yes, I know other also need to be rewrite. And my problem is that `final Predicate<? super NodeInfo> nodeTest = null;`, null is not ok, and how do I make correct `nodeTest`?

**Comment by timurt** [comment]
@HuGanghui 

Please, read default implementation
if you are calling `emptyRootNode.iterateAxis(AxisInfo.DESCENDANT)`, since there are no implementation of this method in `ElementNode` then `NodeInfo.iterateAxis(int axisNumber)` will be called,
If you look at `NodeInfo.iterateAxis(int axisNumber)` you will see
```
    default AxisIterator iterateAxis(int axisNumber) {
        return iterateAxis(axisNumber, AnyNodeTest.getInstance());
    }
```
where `AnyNodeTest` is singleton, thats why you do not have to create `final Predicate<? super NodeInfo> nodeTest = null;`

----

Second, 
```
assertThat("Invalid number of nodes", emptyRootNode.iterateAxis(AxisInfo.DESCENDANT),
                equalTo(new Navigator.AxisFilter(EmptyIterator.ofNodes(), nodeTest)));
```
will not work, because `new Navigator.AxisFilter(EmptyIterator.ofNodes(), nodeTest))` creates new instance and `Navigator.AxisFilter` has no custom `equals` method implementation. Additionally you have no access to `Navigator.AxisFilter` private fields. JUnit will compare them by reference and will be always false.

Instead, you have to manually validate resulting iterator or write `AxisFilter` by yourself with data encapsulation.
I prefer first approach, so your test can be rewritten in the following way:

```
@Test
    public void testRootWithNullDetailAst() {
        final RootNode emptyRootNode = new RootNode(null);
        assertThat("Empty node should not have children", emptyRootNode.hasChildNodes(),
                equalTo(false));

        assertEmptyAxisFilter(emptyRootNode.iterateAxis(AxisInfo.DESCENDANT));
        assertEmptyAxisFilter(emptyRootNode.iterateAxis(AxisInfo.CHILD));
    }

    private void assertEmptyAxisFilter(AxisIterator actualAxisIterator) {
        assertNotNull(actualAxisIterator, "Resulting iterator can not be null");
        assertTrue(actualAxisIterator instanceof Navigator.AxisFilter, "Iterator should be instance of AxisFilter");
        final Navigator.AxisFilter axisFilter = (Navigator.AxisFilter) actualAxisIterator;
        assertNull(axisFilter.next(), "Iterator should be empty");
    }
```

**Comment by romani** [comment]
will be done in scope of https://github.com/checkstyle/checkstyle/pull/8797

**Pull request closed without merging**